bundles/org.openhab.binding.solarman/src/main/java/org/openhab/binding/solarman/internal/typeprovider/SolarmanChannelTypeProvider.java

```diff
@@ -0,0 +1,140 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.solarman.internal.typeprovider;
+
+import static org.openhab.binding.solarman.internal.typeprovider.ChannelUtils.getItemType;
+
+import java.math.BigDecimal;
+import java.net.URI;
+import java.net.URL;
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.solarman.internal.DefinitionParser;
+import org.openhab.binding.solarman.internal.defmodel.InverterDefinition;
+import org.openhab.binding.solarman.internal.defmodel.ParameterItem;
+import org.openhab.core.thing.type.ChannelType;
+import org.openhab.core.thing.type.ChannelTypeBuilder;
+import org.openhab.core.thing.type.ChannelTypeProvider;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.thing.type.StateChannelTypeBuilder;
+import org.openhab.core.types.StateDescriptionFragmentBuilder;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Catalin Sanda - Initial contribution
+ */
+@Component(service = { ChannelTypeProvider.class, SolarmanChannelTypeProvider.class })
+@NonNullByDefault
+public class SolarmanChannelTypeProvider implements ChannelTypeProvider {
+    private final Logger logger = LoggerFactory.getLogger(SolarmanChannelTypeProvider.class);
+    private static final DefinitionParser DEFINITION_PARSER = new DefinitionParser();
+    private static final Pattern INVERTER_DEFINITION_PATTERN = Pattern.compile("/definitions/([^.]+)\\.yaml");
+    private final Map<ChannelTypeUID, ChannelType> channelTypeMap = new ConcurrentHashMap<>();
+
+    @Activate
+    public SolarmanChannelTypeProvider(BundleContext bundleContext) {
+        Collections.list(bundleContext.getBundle().findEntries("/definitions", "*", false)).stream().map(URL::getFile)
+                .map(this::extractInverterDefinitionId).filter(Optional::isPresent).map(Optional::get)
+                .map(this::parseInverterDefinition).forEach(channelTypeMap::putAll);
+    }
+
+    private Map<ChannelTypeUID, ChannelType> parseInverterDefinition(String inverterDefinitionId) {
+        InverterDefinition inverterDefinition = DEFINITION_PARSER.parseDefinition(inverterDefinitionId);
+
+        if (inverterDefinition == null) {
+            logger.error("Unable to parse inverter definition");
```

Comment:
As the Thing/Binding fails, it does not affect openHAB as a whole.
```suggestion
            logger.warn("Unable to parse inverter definition");
```